src/pgen/cluster/agn_triggering.cpp

```diff
@@ -0,0 +1,626 @@
+          auto hydro_pkg = pmb->packages.Get("Hydro");
+          const auto &agn_triggering = hydro_pkg->Param<AGNTriggering>("agn_triggering");
+          return agn_triggering.GetAccretionRate(hydro_pkg.get());
+        },
+        "agn_accretion_rate"));
+  }
+  hydro_pkg->UpdateParam(parthenon::hist_param_key, hst_vars);
+
+  hydro_pkg->AddParam<AGNTriggering>("agn_triggering", *this);
+}
+
+// Compute Cold gas accretion rate within the accretion radius for cold gas triggering
+// and simultaneously remove cold gas (updating conserveds and primitives)
+template <typename EOS>
+void AGNTriggering::ReduceColdMass(parthenon::Real &cold_mass,
+                                   parthenon::MeshData<parthenon::Real> *md,
+                                   const parthenon::Real dt, const EOS eos) const {
+  using parthenon::IndexDomain;
+  using parthenon::IndexRange;
+  using parthenon::Real;
+
+  auto hydro_pkg = md->GetBlockData(0)->GetBlockPointer()->packages.Get("Hydro");
+
+  // Grab some necessary variables
+  const auto &prim_pack = md->PackVariables(std::vector<std::string>{"prim"});
+  const auto &cons_pack = md->PackVariables(std::vector<std::string>{"cons"});
+  IndexRange ib = md->GetBlockData(0)->GetBoundsI(IndexDomain::interior);
+  IndexRange jb = md->GetBlockData(0)->GetBoundsJ(IndexDomain::interior);
+  IndexRange kb = md->GetBlockData(0)->GetBoundsK(IndexDomain::interior);
+  const auto nhydro = hydro_pkg->Param<int>("nhydro");
+  const auto nscalars = hydro_pkg->Param<int>("nscalars");
+
+  const Real accretion_radius2 = pow(accretion_radius_, 2);
+
+  // Reduce just the cold gas
+  const auto units = hydro_pkg->Param<Units>("units");
+  const Real mean_molecular_mass_by_kb = mean_molecular_mass_ / units.k_boltzmann();
+
+  const Real cold_temp_thresh = cold_temp_thresh_;
+  const Real cold_t_acc = cold_t_acc_;
+
+  const bool remove_accreted_mass = remove_accreted_mass_;
+
+  Real md_cold_mass = 0;
+
+  parthenon::par_reduce(
+      parthenon::loop_pattern_mdrange_tag, "AGNTriggering::ReduceColdGas",
+      parthenon::DevExecSpace(), 0, cons_pack.GetDim(5) - 1, kb.s, kb.e, jb.s, jb.e, ib.s,
+      ib.e,
+      KOKKOS_LAMBDA(const int &b, const int &k, const int &j, const int &i,
+                    Real &team_cold_mass) {
+        auto &cons = cons_pack(b);
+        auto &prim = prim_pack(b);
+        const auto &coords = cons_pack.GetCoords(b);
+
+        const parthenon::Real r2 =
+            pow(coords.Xc<1>(i), 2) + pow(coords.Xc<2>(j), 2) + pow(coords.Xc<3>(k), 2);
+        if (r2 < accretion_radius2) {
+
+          const Real temp =
+              mean_molecular_mass_by_kb * prim(IPR, k, j, i) / prim(IDN, k, j, i);
+
+          if (temp <= cold_temp_thresh) {
+
+            const Real cell_cold_mass = prim(IDN, k, j, i) * coords.CellVolume(k, j, i);
+            team_cold_mass += cell_cold_mass;
+
+            const Real cell_delta_rho = -prim(IDN, k, j, i) / cold_t_acc * dt;
```

Comment:
We should double check what `dt` is here. Depending on how it's called we might need to use `beta_dt`

Reply:
Yes, correct.